src/nanoarrow/ipc/writer.c

```diff
@@ -150,3 +155,173 @@ ArrowErrorCode ArrowIpcOutputStreamInitFile(struct ArrowIpcOutputStream* stream,
+    ArrowBufferReset(&private->body_buffer);
+
+    ArrowFree(private);
+  }
+  memset(writer, 0, sizeof(struct ArrowIpcWriter));
+}
+
+static struct ArrowBufferView ArrowBufferToBufferView(const struct ArrowBuffer* buffer) {
+  struct ArrowBufferView buffer_view = {
+      .data.as_uint8 = buffer->data,
+      .size_bytes = buffer->size_bytes,
+  };
+  return buffer_view;
+}
+
+// Eventually, it may be necessary to construct an ArrowIpcWriter which doesn't rely on
+// blocking writes (ArrowIpcOutputStreamWrite). For example an ArrowIpcOutputStream
+// might wrap a socket which is not always able to transmit all bytes of a Message. In
+// that case users of ArrowIpcWriter might prefer to do other work until a socket is
+// ready rather than blocking, or timeout, or otherwise respond to partial transmission.
+//
+// This could be handled by:
+// - keeping partially sent buffers internal and signalling incomplete transmission by
+//   raising EAGAIN, returning "bytes actually written", ...
+//   - when the caller is ready to try again, call ArrowIpcWriterWriteSome()
+// - exposing internal buffers which have not been completely sent, deferring
+//   follow-up transmission to the caller
+
+ArrowErrorCode ArrowIpcWriterWriteSchema(struct ArrowIpcWriter* writer,
+                                         const struct ArrowSchema* in,
+                                         struct ArrowError* error) {
+  NANOARROW_DCHECK(writer != NULL && writer->private_data != NULL && in != NULL);
+  struct ArrowIpcWriterPrivate* private =
+      (struct ArrowIpcWriterPrivate*)writer->private_data;
+
+  NANOARROW_ASSERT_OK(ArrowBufferResize(&private->buffer, 0, 0));
+  NANOARROW_RETURN_NOT_OK(ArrowIpcEncoderEncodeSchema(&private->encoder, in, error));
+  NANOARROW_RETURN_NOT_OK_WITH_ERROR(
+      ArrowIpcEncoderFinalizeBuffer(&private->encoder, /*encapsulate=*/1,
+                                    &private->buffer),
+      error);
+
+  return ArrowIpcOutputStreamWrite(&private->output_stream,
+                                   ArrowBufferToBufferView(&private->buffer), error);
+}
+
+ArrowErrorCode ArrowIpcWriterWriteArrayView(struct ArrowIpcWriter* writer,
+                                            const struct ArrowArrayView* in,
+                                            struct ArrowError* error) {
+  NANOARROW_DCHECK(writer != NULL && writer->private_data != NULL);
+  struct ArrowIpcWriterPrivate* private =
+      (struct ArrowIpcWriterPrivate*)writer->private_data;
+
+  if (in == NULL) {
+    int32_t eos[] = {-1, 0};
+    struct ArrowBufferView data = {.data.as_int32 = eos, .size_bytes = sizeof(eos)};
+    return ArrowIpcOutputStreamWrite(&private->output_stream, data, error);
+  }
+
+  NANOARROW_ASSERT_OK(ArrowBufferResize(&private->buffer, 0, 0));
+  NANOARROW_ASSERT_OK(ArrowBufferResize(&private->body_buffer, 0, 0));
+
+  NANOARROW_RETURN_NOT_OK(ArrowIpcEncoderEncodeSimpleRecordBatch(
+      &private->encoder, in, &private->body_buffer, error));
+
```

Comment:
Hello,

Shouldn't `private->buffer` be finalized here? Otherwise the header of a message isn't written to the output stream.

Like this:
```c
   NANOARROW_RETURN_NOT_OK_WITH_ERROR(
       ArrowIpcEncoderFinalizeBuffer(&private->encoder, /*encapsulate=*/1,
                                     &private->buffer),
       error);
```

Reply:
You know... it really should. Why isn't this failing tests?

Reply:
I have the same question :) When I tried to use this API, I noticed the absence of the header.

Reply:
Okay... I'll get a fix as soon as I write a failing test. Thanks!